### DISCUSSION TITLE
How to add dependencies like Boost with CPMAddPackage?

### DISCUSSION THREAD

**Discussion by viniciusferrao** [TheLartians/Greeter#175] [Q&A]
Hello, this is a help call for anyone willing to help. I'm trying to add this template to an existing project but I'm having some issues with `Boost` and `CLI11` dependencies.

I've changed the main `CMakeLists.txt` file with the following content:

```
# ---- Add dependencies via CPM ----
# see https://github.com/TheLartians/CPM.cmake for more info

#set(CPM_USE_LOCAL_PACKAGES ON)
include(cmake/CPM.cmake)

# PackageProject.cmake will be used to make our target installable
CPMAddPackage("gh:TheLartians/PackageProject.cmake@1.8.0")

CPMAddPackage(
  NAME Boost
  VERSION 1.82.0
  GITHUB_REPOSITORY "boostorg/boost"
  GIT_TAG "boost-1.82.0"
)

CPMAddPackage(
  NAME fmt
  GIT_TAG 10.0.0
  GITHUB_REPOSITORY fmtlib/fmt
  OPTIONS "FMT_INSTALL YES" # create an installable target
)

CPMAddPackage(
  NAME CLI11
  VERSION 2.3.2
  GITHUB_REPOSITORY CLIUtils/CLI11
  OPTIONS "CLI11_INSTALL YES"
)

CPMAddPackage(
  NAME spdlog
  VERSION 1.11.0
  GITHUB_REPOSITORY gabime/spdlog
  OPTIONS "SPDLOG_INSTALL YES"
)

CPMAddPackage(
  NAME magic_enum
  VERSION 0.9.3
  GITHUB_REPOSITORY Neargye/magic_enum
  OPTIONS "MAGIC_ENUM_OPT_INSTALL YES"
)

CPMAddPackage("gh:gsl-lite/gsl-lite@0.41.0")
```

And added to `target_link_libraries`:

```
# Link dependencies
target_link_libraries(${PROJECT_NAME} PRIVATE fmt::fmt CLI11::CLI11 Boost::headers Boost::system Boost::thread spdlog::spdlog gsl::gsl-lite magic_enum::magic_enum resolv)
```

But when I try to configure with `cmake -S all -B build` I receive this error message:

```
-- Configuring done
CMake Error: install(EXPORT "CloysterHPCTargets" ...) includes target "CloysterHPC" which requires target "CLI11" that is not in any export set.
CMake Error: install(EXPORT "CloysterHPCTargets" ...) includes target "CloysterHPC" which requires target "boost_headers" that is not in any export set.
CMake Error: install(EXPORT "CloysterHPCTargets" ...) includes target "CloysterHPC" which requires target "boost_system" that is not in any export set.
CMake Error: install(EXPORT "CloysterHPCTargets" ...) includes target "CloysterHPC" which requires target "boost_thread" that is not in any export set.
-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

I based the Boost settings from https://github.com/cpm-cmake/CPM.cmake (the snippets in `README.md`). Everything else I tried to figure out by myself.

Also I'm not sure if I done the changes in the right place, I don't get if I should add the required dependencies to the main CMakeLists.txt or if I should use the `standalone` directory and change `CMakeLists.txt` there. I didn't understand it's purpose correctly to be honest.

Thanks.

**Comment by ClausKlein** [comment]
You have to install all your dependent packages!

**Reply by viniciusferrao** [reply]
You mean by setting the INSTALL option? I check for example the source code of CLI11 and found `OPTIONS "CLI11_INSTALL YES"`, so I have enabled it on CPMAddPackage. Is that what you're referring to?

**Reply by ClausKlein** [reply]
Yes

**Reply by viniciusferrao** [reply]
Well something should be broken with the other projects, because that's what I've done and it's also on the `CMakeLists.txt` snipped that I showed. :(

**Reply by ClausKlein** [reply]
What is about boost, it seems not to have an install rule?

You may also try to disable the EXCLUDE_FROM_ALL CPM.cmake option.
